util/reggen/gen_rust.py

```diff
@@ -0,0 +1,394 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+"""
+Generate Rust constants from validated register JSON tree
+"""
+
+import io
+import logging as log
+import sys
+import textwrap
+import warnings
+from typing import List, Optional, Set, TextIO
+
+
+from .field import Field
+from .ip_block import IpBlock
+from .params import LocalParam
+from .register import Register
+from .multi_register import MultiRegister
+from .signal import Signal
+from .window import Window
+
+
+def genout(outfile: TextIO, msg: str) -> None:
+    outfile.write(msg)
+
+def to_snake_case(s: str) -> str:
+    val = []
+    for i, ch in enumerate(s):
+      if i > 0 and ch.isupper():
+        val.append('_')
+      val.append(ch)
+    return ''.join(val)
+
+def as_define(s: str) -> str:
+    s = s.upper()
+    r = ''
+    for i in range(0, len(s)):
+        r += s[i] if s[i].isalnum() else '_'
+    return r
+
+
+def first_line(s: str) -> str:
+    """Returns the first line of a multi-line string"""
+    return s.splitlines()[0]
+
+
+def format_comment(s: str) -> str:
+    """Formats a string to comment wrapped to an 80 character line width
+
+    Returns wrapped string including newline and // comment characters.
+    """
+    return '\n'.join(
+        textwrap.wrap(
+            s, width=77, initial_indent='// ', subsequent_indent='// ')) + '\n'
+
+
+def is_negative(x: str) -> bool:
+    """ Checks if a decimal string is a negative value. """
+
+    # Try and parse decimal string.
+    try:
+        n = int(x)
+        return n < 0
+    except ValueError:
+        return False
+
+
+def data_type(name: str, val: str) -> str:
+    """ Returns proper data type for name-value pair. """
+
+    if name.endswith("_OFFSET") or name.endswith("_BASE_ADDR"):
+        return "usize"
+    elif is_negative(val):
+        return "i32"
+    else:
+        return "u32"
```

Comment:
I guess we don't have any examples where this would be a problem, but (once you're passing an actual integer to this function) it might be worth checking that it fits in 32 bits? It's probably nicer to spit out an error as part of the regtool call than sometime later when compiling the auto-generated code.

Reply:
Done.